dao/src/lib.rs

```diff
@@ -0,0 +1,692 @@
+        if self.proposal_id > 0 {
+            let previous_starting_period = self
+                .proposals
+                .get(&(&self.proposal_id - 1))
+                .unwrap()
+                .starting_period;
+            if starting_period < previous_starting_period + self.period_duration {
+                starting_period = previous_starting_period + self.period_duration;
+            }
+        }
+
+        let proposal = Proposal {
+            proposer: msg::source(),
+            applicant: *applicant,
+            quorum,
+            amount,
+            details,
+            starting_period,
+            ..Proposal::default()
+        };
+
+        self.proposals.insert(self.proposal_id, proposal);
+
+        msg::reply(
+            DaoEvent::SubmitFundingProposal {
+                proposer: msg::source(),
+                applicant: *applicant,
+                proposal_id: self.proposal_id,
+                amount,
+            },
+            0,
+        );
+        self.proposal_id = self.proposal_id.saturating_add(1);
+    }
+
+    // The member (or the delegate address of the member) submit his vote (YES or NO) on the proposal
+    // Requirements:
+    // * The proposal can be submitted only by the existing members or their delegate addresses
+    // * The member can vote on the proposal only once
+    // * Proposal must exist, the voting period must has started and not expired
+    // Arguments:
+    // * `proposal_id`: the proposal ID
+    // * `vote`: the member  a member vote (YES or NO)
+    fn submit_vote(&mut self, proposal_id: u128, vote: Vote) {
+        // check that `msg::source()` is either a DAO member or a delegate key
+        match self.member_by_delegate_key.get(&msg::source()) {
+            Some(member) => {
+                if !self.is_member(member) {
+                    panic!("account is not a DAO member");
+                }
+            }
+            None => {
+                panic!("account is not a delegate");
+            }
+        }
+
+        // checks that proposal exists, the voting period has started, not expired and that member did not vote on the proposal
+        match self.proposals.get(&proposal_id) {
+            Some(proposal) => {
+                if exec::block_timestamp() > proposal.starting_period + self.voting_period_length {
+                    panic!("proposal voting period has expired");
+                }
+                if exec::block_timestamp() < proposal.starting_period {
+                    panic!("voting period has not started");
+                }
+                if proposal.votes_by_member.contains_key(&msg::source()) {
+                    panic!("account has already voted on that proposal");
+                }
+            }
+            None => {
+                panic!("proposal does not exist");
+            }
+        }
+
+        let proposal = self.proposals.get_mut(&proposal_id).unwrap();
```

Comment:
Double getting is suboptimal.
```suggestion
        let proposal = match self.proposals.get_mut(&proposal_id) {
            Some(proposal) => {
                if exec::block_timestamp() > proposal.starting_period + self.voting_period_length {
                    panic!("proposal voting period has expired");
                }
                if exec::block_timestamp() < proposal.starting_period {
                    panic!("voting period has not started");
                }
                if proposal.votes_by_member.contains_key(&msg::source()) {
                    panic!("account has already voted on that proposal");
                }
                proposal
            }
            None => {
                panic!("proposal does not exist");
            }
        }
```